src/caches/beliefpropagationcache.jl

```diff
@@ -14,12 +14,20 @@ end
   return default_bp_maxiter(undirected_graph(underlying_graph(g)))
 end
 default_partitioned_vertices(ψ::AbstractITensorNetwork) = group(v -> v, vertices(ψ))
+function default_partitioned_vertices(f::AbstractFormNetwork)
+  return group(v -> original_state_vertex(f, v), vertices(f))
+end
 default_cache_update_kwargs(cache) = (; maxiter=20, tol=1e-5)
+function default_cache_construction_kwargs(alg::Algorithm"bp", ψ::AbstractITensorNetwork)
+  return (; partitioned_vertices=default_partitioned_vertices(ψ))
+end
 
+#TODO: Define a version of this that works for QN supporting tensors
 function message_diff(message_a::Vector{ITensor}, message_b::Vector{ITensor})
   lhs, rhs = contract(message_a), contract(message_b)
-  return 0.5 *
-         norm((denseblocks(lhs) / sum(diag(lhs))) - (denseblocks(rhs) / sum(diag(rhs))))
+  tr_lhs = length(inds(lhs)) == 1 ? sum(lhs) : sum(diag(lhs))
+  tr_rhs = length(inds(rhs)) == 1 ? sum(rhs) : sum(diag(rhs))
+  return 0.5 * norm((denseblocks(lhs) / tr_lhs) - (denseblocks(rhs) / tr_rhs))
```

Comment:
Could you explain this?

Reply:
If the message has only one index then `diag(A::ITensor)` is not defined and doesn't work - which I am not sure if it is intentional or not? Technically I might imagine it should return a vector of the elements of the `ITensor` if it has only one index

Reply:
That's just how `diag` is defined in Julia:
```julia
julia> using LinearAlgebra

julia> diag(randn(2, 2))
2-element Vector{Float64}:
 -0.7523337727717532
  1.1916982630642805

julia> diag(randn(2))
ERROR: ArgumentError: use diagm instead of diag to construct a diagonal matrix
Stacktrace:
 [1] diag(A::Vector{Float64})
   @ LinearAlgebra ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/generic.jl:445
 [2] top-level scope
   @ REPL[11]:1

julia> diag(randn(2, 2, 2))
ERROR: MethodError: no method matching diag(::Array{Float64, 3})

Closest candidates are:
  diag(::BitMatrix)
   @ LinearAlgebra ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/bitarray.jl:79
  diag(::ITensors.ITensor)
   @ ITensors ~/.julia/packages/ITensors/c1wvC/src/itensor.jl:1902
  diag(::UnitLowerTriangular)
   @ LinearAlgebra ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/triangular.jl:418
  ...

Stacktrace:
 [1] top-level scope
   @ REPL[12]:1
```
but I agree that there should be a generic function, maybe called `diagonals`, for any array order that works as you say (I have a function `diagview` in `DiagonalArrays` that acts like that: https://github.com/ITensor/ITensors.jl/tree/v0.3.66/NDTensors/src/lib/DiagonalArrays).

But what happens if the message tensors have more than two indices? How should `message_diff` be defined? I think I'm just missing the broader picture/goal for this function.

Reply:
Any thoughts on this? Or do we leave as is and remind ourselves to come up with a better definition that works more streamlined with QN tensors and one index tensors.

Reply:
It seems like a better design would be to allow customization of a norm for the message tensors. `tr` is just one choice, which I guess is good if it is a norm network but otherwise normalizing by `norm` of the message tensors may be better (and work for QNs).

Reply:
Also I'm sure we can fix issues with QNs in a naive way by just writing a specialized trace function as a loop over the diagonal values, which would also simplify this code.

Reply:
Okay let me come up with something that attempts to unify this and relies on the ability to pass a notion of `norm` as a keyword argument to the message diff function

Reply:
Okay I refactored this so `message_diff()` excepts a keyword argument for how to compute the `message_norm` which defaults to `norm()` thus working for QNS and ITensors of dimension 1 in the default case.

Seems to be working well. Let me know what you think.

Reply:
Looks good, thanks! We could always bring back the previous trace-based norm as needed.